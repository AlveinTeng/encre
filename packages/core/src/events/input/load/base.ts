import { Context } from './context';

/**
 * Interface that defines the events for loading readable contexts.
 */
export interface Loader {
  load(): Promise<Context[]>;
}

/**
 * Abstract class that provides an abstract load() event from the
 * Loader interface.
 */
export abstract class BaseLoader implements Loader {
  /**
   * Abstract method that loads the readable contexts.
   * @returns A Promise that resolves with an array of Context instances.
   */
  abstract load(): Promise<Context[]>;
}
