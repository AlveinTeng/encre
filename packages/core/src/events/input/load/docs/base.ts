import { BaseEvent, BaseEventParams } from '../../../base.js';
import { Context } from './context.js';

/**
 * Interface that defines the loader for loading readable source.
 */
export interface DocLoader<Input = unknown, Output = Context[]> {
  load(source: Input): Promise<Output>;
}

export interface BaseDocLoaderCallOptions extends BaseEventParams {
  /**
   * Whether the loader should split the context into chunks
   */
  shouldSplit?: boolean;
}

export interface BaseLoaderParams extends BaseDocLoaderCallOptions {}

/**
 * Abstract class that provides an abstract load() event from the Loader
 * interface.
 */
export abstract class BaseLoader<
    CallInput = unknown,
    CallOutput = Context[],
    CallOptions extends BaseDocLoaderCallOptions = BaseDocLoaderCallOptions,
  >
  extends BaseEvent<CallInput, CallOutput, CallOptions>
  implements DocLoader<CallInput, CallOutput>
{
  declare CallOptions: CallOptions;

  _namespace: string[] = ['input', 'load', 'docs'];

  shouldSplit?: boolean;

  constructor(fields?: BaseLoaderParams) {
    super(fields ?? {});

    this.shouldSplit = fields?.shouldSplit ?? false;
  }

  async invoke(
    input: CallInput,
    options?: Partial<CallOptions>
  ): Promise<CallOutput> {
    this.shouldSplit = options?.shouldSplit ?? this.shouldSplit;

    return this.load(input);
  }

  /**
   * Abstract method that loads the readable source.
   * @returns A Promise that resolves with an array of `Context` instances.
   */
  abstract load(source: CallInput): Promise<CallOutput>;
}
