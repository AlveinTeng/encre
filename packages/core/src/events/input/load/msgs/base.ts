import {
  Serializable,
  SerializedConstructor,
} from '../../../../load/serializable.js';
import {
  SerializedMessage,
  SerializedMessageData,
} from '../../../output/provide/message.js';

export type MessageRole =
  | 'human'
  | 'assistant'
  | 'system'
  | 'function'
  | 'general';

export type Message = {
  content: string;
  role?: MessageRole;
  name?: string;
};

export type BaseMessageLike = BaseMessage | [MessageRole, string] | string;

/**
 * Interface that defines the fields in `BaseMessage`.
 */
export interface BaseMessageFields {
  content: string;
  name?: string;
  additionalKwargs?: {
    [key: string]: unknown;
  };
}

/**
 * Abstract class that stores the message from the `BaseMessageFields`
 * interface.
 */
export abstract class BaseMessage
  extends Serializable
  implements BaseMessageFields
{
  _namespace: string[] = ['input', 'load', 'msgs'];

  content: string;

  name?: string;

  additionalKwargs?: NonNullable<BaseMessageFields['additionalKwargs']>;

  abstract _role(): MessageRole;

  constructor(fields: string | BaseMessageFields) {
    if (typeof fields === 'string') {
      fields = { content: fields };
    }
    super(fields);

    this.name = fields.name;
    this.content = fields.content;
    this.additionalKwargs = fields.additionalKwargs;
  }

  toSerialized(): SerializedMessage {
    return {
      role: this._role(),
      json: (this.toJSON() as SerializedConstructor)
        ._kwargs as object as SerializedMessageData,
    };
  }

  abstract concat(message: BaseMessage): BaseMessage;

  static _mergeAdditionalKwargs(
    left: NonNullable<BaseMessageFields['additionalKwargs']>,
    right: NonNullable<BaseMessageFields['additionalKwargs']>
  ): NonNullable<BaseMessageFields['additionalKwargs']> {
    const additionalKwargs = { ...left };
    const newKwargs = { ...right };

    for (const [k, v] of Object.entries(newKwargs)) {
      if (typeof additionalKwargs[k] !== typeof v) {
        throw new Error(
          `key ${k} in additionalKwargs but values do not match types.`
        );
      } else if (additionalKwargs[k] === undefined) {
        additionalKwargs[k] = v;
      } else if (typeof additionalKwargs[k] === 'string') {
        additionalKwargs[k] = (additionalKwargs[k] as string) + v;
      } else if (
        !Array.isArray(additionalKwargs[k]) &&
        typeof additionalKwargs[k] === 'object'
      ) {
        additionalKwargs[k] = this._mergeAdditionalKwargs(
          additionalKwargs[k] as NonNullable<
            BaseMessageFields['additionalKwargs']
          >,
          v as NonNullable<BaseMessageFields['additionalKwargs']>
        );
      } else {
        throw new Error(
          `key ${k} already in additionalKwargs but failed to merge.`
        );
      }
    }

    return additionalKwargs;
  }
}

export interface ChatMessageFields extends BaseMessageFields {
  role: string;
}

export class ChatMessage extends BaseMessage {
  _isSerializable = true;

  static _name(): string {
    return 'ChatMessage';
  }

  role: string;

  constructor(content: string, role: string);

  constructor(fields: ChatMessageFields);

  constructor(fields: string | ChatMessageFields, role?: string) {
    if (typeof fields === 'string') {
      fields = { content: fields, role: role! as MessageRole };
    }
    super(fields);

    this.role = fields.role;
  }

  _role(): MessageRole {
    return 'general';
  }

  concat(message: ChatMessage): ChatMessage {
    return new ChatMessage({
      content: this.content + message.content,
      additionalKwargs: ChatMessage._mergeAdditionalKwargs(
        this.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >,
        message.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >
      ),
      role: this.role,
    });
  }

  static isInstance(message: BaseMessage): message is ChatMessage {
    return message._role() === 'general';
  }
}

/**
 * Class that stores the Human message in a conversation.
 */
export class HumanMessage extends BaseMessage {
  _isSerializable = true;

  _role(): MessageRole {
    return 'human';
  }

  static _name(): string {
    return 'HumanMessage';
  }

  concat(message: HumanMessage): HumanMessage {
    return new HumanMessage({
      content: this.content + message.content,
      additionalKwargs: HumanMessage._mergeAdditionalKwargs(
        this.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >,
        message.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >
      ),
    });
  }
}

/**
 * Class that stores the Bot message in a conversation.
 */
export class BotMessage extends BaseMessage {
  _isSerializable = true;

  _role(): MessageRole {
    return 'assistant';
  }

  static _name(): string {
    return 'BotMessage';
  }

  concat(message: BotMessage): BotMessage {
    return new BotMessage({
      content: this.content + message.content,
      additionalKwargs: BotMessage._mergeAdditionalKwargs(
        this.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >,
        message.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >
      ),
    });
  }
}

/**
 * Class that stores the System message in a conversation.
 */
export class SystemMessage extends BaseMessage {
  _isSerializable = true;

  _role(): MessageRole {
    return 'system';
  }

  static _name(): string {
    return 'SystemMessage';
  }

  concat(message: SystemMessage): SystemMessage {
    return new SystemMessage({
      content: this.content + message.content,
      additionalKwargs: SystemMessage._mergeAdditionalKwargs(
        this.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >,
        message.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >
      ),
    });
  }
}

/**
 * Class that stores the Function message in a conversation.
 */
export class FunctionMessage extends BaseMessage {
  _isSerializable = true;

  _role(): MessageRole {
    return 'function';
  }

  static _name(): string {
    return 'FunctionMessage';
  }

  concat(message: FunctionMessage): FunctionMessage {
    return new FunctionMessage({
      content: this.content + message.content,
      additionalKwargs: FunctionMessage._mergeAdditionalKwargs(
        this.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >,
        message.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >
      ),
      name: this.name ?? '',
    });
  }
}

export function getChatString(
  messages: BaseMessage[],
  replaceIdentityByName: boolean = false,
  humanIndentity: string = 'Human',
  botIdentity: string = 'AI'
): string {
  const messagesStr: string[] = [];
  for (const msg of messages) {
    let role: string;
    if (msg._role() === 'human') {
      role = humanIndentity;
    } else if (msg._role() === 'assistant') {
      role = botIdentity;
    } else if (msg._role() === 'system') {
      role = 'System';
    } else if (msg._role() === 'function') {
      role = 'Function';
    } else if (ChatMessage.isInstance(msg)) {
      role = msg.role;
    } else {
      throw new Error(
        `Current message role unsupported: ${JSON.stringify(msg)}`
      );
    }

    if (replaceIdentityByName && msg.name) {
      role = msg.name;
    }

    messagesStr.push(`${role}: ${msg.content}`);
  }

  return messagesStr.join('\n');
}

export function convertMessageLikeToMessage(
  messageLike: BaseMessageLike
): BaseMessage {
  if (typeof messageLike === 'string') {
    return new HumanMessage(messageLike);
  } else if (isBaseMessage(messageLike)) {
    return messageLike;
  }

  const [role, content] = messageLike;
  if (role === 'human') {
    return new HumanMessage({ content });
  } else if (role === 'assistant') {
    return new BotMessage({ content });
  } else if (role === 'system') {
    return new SystemMessage({ content });
  } else {
    throw new Error('CANNOT convert MessageLike to Message');
  }
}

export function isBaseMessage(
  messageLike: unknown
): messageLike is BaseMessage {
  return 'content' in (messageLike as BaseMessage);
}
