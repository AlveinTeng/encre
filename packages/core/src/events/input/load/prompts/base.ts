import { Serializable } from '../../../../load/serializable.js';
import { CallableConfig } from '../../../../record/callable.js';
import { VariableValidator } from '../../../../utils/promptTemplateValidator/variableValidator.js';
import { BaseEvent, BaseEventParams } from '../../../base.js';
import { BaseMessage } from '../msgs/base.js';

export abstract class BasePrompt extends Serializable {
  _isSerializable = false;

  _namespace: string[] = ['input', 'load', 'prompts', this._promptType()];

  abstract _promptType(): string;

  abstract toString(): string;

  abstract toChatMessages(): BaseMessage[];
}

export interface PromptTemplateParams extends BaseEventParams {
  /**
   * the template string
   */
  template: string;

  /**
   * Clearly declare the input variables inside the template
   */
  inputVariables: string[];

  partialVariables: string[];

  validator?: VariableValidator;
}

export interface BasePromptTemplateInput {
  [key: string]: unknown;
};

export abstract class BasePromptTemplate<
    CallInput extends BasePromptTemplateInput = BasePromptTemplateInput,
    CallOutput extends BasePrompt = BasePrompt,
    CallOptions extends CallableConfig = CallableConfig,
  >
  extends BaseEvent<CallInput, CallOutput, CallOptions>
  implements PromptTemplateParams
{
  _namespace: string[] = ['input', 'load', 'prompts', this._templateType()];

  static _name(): string {
    return 'BasePromptTemplate';
  }

  template = '';

  inputVariables: string[] = [];

  partialVariables: string[] = [];

  validator?: VariableValidator;

  constructor(fields?: Partial<PromptTemplateParams>) {
    super(fields ?? {});

    this.template = fields?.template ?? this.template;

    this.inputVariables = fields?.inputVariables ?? this.inputVariables;
    this.partialVariables = fields?.partialVariables ?? this.partialVariables;

    this.validator = fields?.validator ?? this.validator;

    if (this.template && this.inputVariables) {
      this._isInputExists(this.template, this.inputVariables);
    }
  }

  abstract _templateType(): string;

  /**
   * To validate whether the inputVariables is valid
   * @param template
   * @param inputVariables
   */
  private _isInputExists(template: string, inputVariables: string[]): void {
    const variablePattern = /\{\{([^}]+)\}\}/g;
    let match: RegExpExecArray | null;
    const foundVariables: Set<string> = new Set();

    while ((match = variablePattern.exec(template)) !== null) {
      foundVariables.add(match[1]);
    }

    inputVariables.forEach((variable) => {
      if (!foundVariables.has(variable)) {
        throw new Error(
          `Variable '${variable}' is declared but not used in the template.`
        );
      }
    });
  }

  async invoke(
    input: CallInput,
    options?: Partial<CallOptions>
  ): Promise<CallOutput> {
    if (!this.validate(input)) {
      throw new Error('the validation for inputValue failed');
    }

    return this.formatPrompt(input);
  }

  validate(input: CallInput): boolean {
    if (this.validator) {
      const validationResult = this.validator.validate(input);
      if (!validationResult.isValid) {
        return false;
      }
    }
    return true;
  }

  abstract format(input: CallInput): Promise<string>;

  abstract formatPrompt(input: CallInput): Promise<CallOutput>;
}
