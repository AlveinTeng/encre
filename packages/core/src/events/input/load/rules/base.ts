import { RecordId } from '../../../../load/keymap';
import { Serializable } from '../../../../load/serializable';
import { SerializedRuleCollection } from '../../../../studio/serde';
import { BaseRule } from '../../../inference/validate/guardrails/base';

export interface BaseRuleCollectionField {
  collection: Record<string, BaseRule | BaseRuleCollection>;
  conjunction?: 'and' | 'or'; // default is AND
}

export class BaseRuleCollection
  extends Serializable
  implements BaseRuleCollectionField
{
  _namespace: string[] = ['events', 'input', 'load', 'rules'];

  collection: Record<string, BaseRule | BaseRuleCollection>;

  conjunction: 'and' | 'or';

  constructor(fields: BaseRuleCollectionField) {
    super(fields);

    this.collection = fields.collection;
    this.conjunction = fields.conjunction ?? 'and';
  }

  get description(): string {
    return Object.entries(this.collection)
      .map(([subject, rule]) => {
        const currDescription: string = rule.description;
        const isInvert: boolean = this._isInvertLogic(
          currDescription,
          this.conjunction
        );
        const isCollection: boolean = rule instanceof BaseRuleCollection;

        return isInvert
          ? `(${
              !isCollection ? `{{{${subject}}}} ` : ''
            }<${subject}>${currDescription}</${subject}>)`
          : `${
              !isCollection ? `{{{${subject}}}} ` : ''
            }<${subject}>${currDescription}</${subject}>`;
      })
      .join(` ${this.conjunction.toUpperCase()} `);
  }

  static async deserialize(
    serialized: SerializedRuleCollection,
    values: Record<string, unknown> = {}
  ): Promise<BaseRuleCollection> {
    if (serialized._type !== 'rule-collection') {
      throw new Error(
        `CANNOT deserialize this type in rule collection: ${serialized._type}`
      );
    }

    const { BaseRule } = await import(
      '../../../inference/validate/guardrails/base.js'
    );

    const ruleCollectionField: BaseRuleCollectionField = {
      collection: Object.fromEntries(
        await Promise.all(
          Object.entries(serialized.collection).map(async ([subject, rule]) => [
            subject,
            rule._type === 'rule-collection'
              ? await BaseRuleCollection.deserialize(
                  rule,
                  (values?.[subject] ?? {}) as Record<string, unknown>
                )
              : await BaseRule.deserialize(
                  rule,
                  (values?.[subject] ?? {}) as Record<string, unknown>
                ),
          ])
        )
      ),
      conjunction: serialized.conjunction,
    };

    return new BaseRuleCollection(ruleCollectionField);
  }

  serialize(): SerializedRuleCollection {
    return {
      _type: 'rule-collection',
      description: this.description,
      collection: Object.fromEntries(
        Object.entries(this.collection).map(([subject, rule]) => [
          subject,
          rule.serialize(),
        ])
      ),
      conjunction: this.conjunction,
    };
  }

  getCleanDescription(): string {
    return Object.entries(this.collection)
      .map(([subject, rule]) => {
        const currDescription: string = rule.getCleanDescription();
        const isInvert: boolean = this._isInvertLogic(
          currDescription,
          this.conjunction
        );
        const isCollection: boolean = rule instanceof BaseRuleCollection;

        return isInvert
          ? `(${!isCollection ? `${subject} ` : ''}${currDescription})`
          : `${!isCollection ? `${subject} ` : ''}${currDescription}`;
      })
      .join(` ${this.conjunction.toUpperCase()} `);
  }

  async validate(
    input: {
      [key in string]: unknown;
    },
    variables?: {
      [key in string]?: Record<string, unknown>;
    }
  ): Promise<boolean> {
    if (
      !this._isIdenticalKeySet(Object.keys(this.collection), Object.keys(input))
    ) {
      throw new Error(
        'CANNOT perform validation in rule collection because of two different key sets.'
      );
    }

    const recordHistory: Set<RecordId> = new Set([this._recordId]);

    if (Object.keys(this.collection).length === 0) {
      return false;
    }

    if (this.conjunction === 'and') {
      return Object.entries(this.collection).every(async ([subject, rule]) => {
        return this._validate(subject, rule, recordHistory, input, variables);
      });
    }

    return Object.entries(this.collection).some(async ([subject, rule]) => {
      return this._validate(subject, rule, recordHistory, input, variables);
    });
  }

  concat(
    ruleCollection: BaseRuleCollection,
    conjunction: 'and' | 'or'
  ): BaseRuleCollection {
    return new BaseRuleCollection({
      collection: {
        '0': this,
        '1': ruleCollection,
      },
      conjunction,
    });
  }

  protected async _validate(
    subject: string,
    rule: BaseRule | BaseRuleCollection,
    recordHistory: Set<RecordId>,
    input: {
      [key in string]: unknown;
    },
    variables?: {
      [key in string]?: Record<string, unknown>;
    }
  ): Promise<boolean> {
    if (rule instanceof BaseRule) {
      return await rule.validate(input[subject], variables?.[subject]);
    }
    const collectionVariables = Object.fromEntries(
      Object.entries(variables ?? {})
        .filter(([cSubject, _]) => {
          return cSubject.startsWith(`${subject}.`);
        })
        .map(([cSubject, cVariables]) => {
          const substring = `${subject}.`;
          const startIndex = cSubject.indexOf(substring);

          if (startIndex === -1) {
            throw new Error(
              `CANNOT perform validation bacause of incorrect key name of the variable: ${subject}`
            );
          }

          const sliceFromIndex = startIndex + substring.length;
          return [cSubject.substring(sliceFromIndex), cVariables];
        })
    );

    if (recordHistory.size === 17) {
      throw new Error('number of collections reaches the maximum');
    }

    if (recordHistory.has(rule._recordId)) {
      throw new Error('CANNOT recursively register rule collection');
    }
    recordHistory.add(rule._recordId);

    return await rule.validate(input, collectionVariables);
  }

  protected _isIdenticalKeySet(keyset1: string[], keyset2: string[]): boolean {
    if (keyset1.length !== keyset2.length) {
      return false;
    }

    const sortedKeys1 = keyset1.slice().sort();
    const sortedKeys2 = keyset1.slice().sort();

    for (let i = 0; i < sortedKeys1.length; i++) {
      if (sortedKeys1[i] !== sortedKeys2[i]) {
        return false;
      }
    }

    return true;
  }

  protected _isCompositeRule(description: string): boolean {
    return description.includes('AND') || description.includes('OR');
  }

  protected _getOutermostConjunction(description: string): string {
    const isMatchingParentheses = (expression: string): boolean => {
      let balance = 0;
      for (let i = 0; i < expression.length; i++) {
        if (expression[i] === '(') balance++;
        else if (expression[i] === ')') balance--;

        if (balance < 0) return false;
      }
      return balance === 0;
    };

    let expression = description.trim();

    while (
      expression.startsWith('(') &&
      expression.endsWith(')') &&
      isMatchingParentheses(expression)
    ) {
      expression = expression.substring(1, expression.length - 1).trim();
    }

    let depth = 0;
    let cleanedExpression = '';

    for (let i = 0; i < expression.length; i++) {
      const char = expression[i];
      if (char === '(') {
        depth++;
      } else if (char === ')') {
        depth--;
      } else if (depth === 0) {
        cleanedExpression += char;
      }
    }

    const andIndex: number = cleanedExpression.indexOf('AND');
    const orIndex: number = cleanedExpression.indexOf('OR');

    if (andIndex !== -1 && (andIndex < orIndex || orIndex === -1)) {
      return 'AND';
    } else if (orIndex !== -1 && (orIndex < andIndex || andIndex === -1)) {
      return 'OR';
    }

    throw new Error(
      `description CANNOT get outmost conjunction: ${description}`
    );
  }

  protected _isInvertLogic(
    description: string,
    conjunction: 'and' | 'or'
  ): boolean {
    if (!this._isCompositeRule(description)) {
      return false;
    }

    const currConjunction: string = this._getOutermostConjunction(description);
    return currConjunction.toLowerCase() !== conjunction;
  }
}
