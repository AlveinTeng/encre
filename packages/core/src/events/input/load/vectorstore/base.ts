import { Serializable } from '../../../../load/serializable.js';
import { Context } from '../docs/context.js';

export interface BaseVectorStoreField {}

export abstract class BaseVectorStore
  extends Serializable
  implements BaseVectorStoreField
{
  declare FilterType: object | string;

  _isSerializable = false;

  _namespace: string[] = [
    'events',
    'input',
    'load',
    'vectorstore',
    this._vectorstoreType(),
  ];

  constructor(fields?: Partial<BaseVectorStoreField>) {
    super(fields);
  }

  abstract _vectorstoreType(): string;

  abstract addVectors(
    embeddings: number[][],
    context: Context[],
    options?: Record<string, unknown>
  ): Promise<void>;

  abstract deleteVectors(params?: Record<string, unknown>): Promise<void>;

  abstract similaritySearch(
    query: number[],
    topK: number,
    filter?: this['FilterType']
  ): Promise<[Context, number][]>;

  abstract maxMarginalRelevanceSearch(
    query: number[],
    topK: number,
    lambda: number,
    filter?: this['FilterType']
  ): Promise<[Context, number][]>;
}
