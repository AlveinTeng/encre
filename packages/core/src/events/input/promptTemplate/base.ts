import { Callable } from "../../../record/callable";
import { baseValidator } from "../../../utils/pValidators/baseValidator";
type PromptTemplateArgs = {
    inputVariables?: string[];
    template: string;
};

export abstract class pTemplateBase {

    private inputVariables: string[];
    private template: string;

    constructor(args: PromptTemplateArgs) {
        this.inputVariables= args.inputVariables? args.inputVariables : [];
        this.template = args.template;
    }

    async format(values: { [key: string]: string }): Promise<string> {
        let formattedTemplate = this.template;
        
        for (const variable of this.inputVariables) {
            const value = values[variable];
            if (value) {
                formattedTemplate = formattedTemplate.replace(`{${variable}}`, value);
            }
        }

        return formattedTemplate;
    }
}

