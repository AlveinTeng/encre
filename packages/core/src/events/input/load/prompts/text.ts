import { BaseMessage, HumanMessage } from '../msgs/base.js';
import {
  BasePrompt,
  BasePromptTemplate,
  BasePromptTemplateInput,
  PromptTemplateParams,
} from './base.js';

export class StringPrompt extends BasePrompt {
  _isSerializable = true;

  static _name(): string {
    return 'StringPrompt';
  }

  value: string;

  constructor(value: string, ...args) {
    super({
      value,
      ...args,
    });
    this.value = value;
  }

  _promptType(): string {
    return 'text';
  }

  toString(): string {
    return this.value;
  }

  toChatMessages(): BaseMessage[] {
    return [new HumanMessage(this.value)];
  }
}

export interface StringPromptTemplateParams extends PromptTemplateParams {
  prefix?: string;
  suffix?: string;
}

export class StringPromptTemplate
  extends BasePromptTemplate<BasePromptTemplateInput, StringPrompt>
  implements StringPromptTemplateParams
{
  _isSerializable = true;

  static _name(): string {
    return 'StringPromptTemplate';
  }

  prefix?: string;

  suffix?: string;

  constructor(fields: Partial<StringPromptTemplateParams>) {
    super(fields);

    this.prefix = fields.prefix;
    this.suffix = fields.suffix;
  }

  _templateType(): string {
    return 'text';
  }

  async format(input: BasePromptTemplateInput): Promise<string> {
    let formattedTemplate: string = this.template;

    // Iterate through each input variable and replace placeholders with actual values
    this.inputVariables?.forEach((variable: string) => {
      if (!Object.prototype.hasOwnProperty.call(input, variable)) {
        // If a required variable is missing in the input, throw an error
        throw new Error(`Missing value for variable '${variable}'`);
      }

      const regex = new RegExp(`\\{\\{${variable}\\}\\}`, 'g');
      formattedTemplate = formattedTemplate.replace(
        regex,
        String(input[variable])
      );
    });

    return [this.prefix ?? '', formattedTemplate, this.suffix ?? ''].join('');
  }

  async formatPrompt(input: BasePromptTemplateInput): Promise<StringPrompt> {
    const formattedPrompt: string = await this.format(input);

    return new StringPrompt(formattedPrompt);
  }
}
