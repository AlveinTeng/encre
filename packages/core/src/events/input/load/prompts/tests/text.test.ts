import { expect, test, describe } from '@jest/globals';
import { stringify } from 'yaml';
import { GeneralRule } from '../../../../inference/validate/guardrails/base.js';
import { StringPrompt, StringPromptTemplate } from '../text.js';

test('test custom StringPrompt', async () => {
  const testStringPrompt = new StringPrompt('this is a test prompt.');

  // StringPrompt is not serializable
  const serializedStr: string = JSON.stringify(testStringPrompt, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  expect(testStringPrompt.toString()).toBe('this is a test prompt.');

  // ChatMessages are serializable
  const serializedStr2: string = JSON.stringify(
    testStringPrompt.toChatMessages(),
    null,
    2
  );
  expect(stringify(JSON.parse(serializedStr2))).toMatchSnapshot();
});

describe('test StringPromptTemplate', () => {
  test('declare not exist input variables', async () => {
    const inputVariables = ['usedVariable', 'unusedVariable'];
    const template = 'This is a template with {{usedVariable}}';

    expect(() => {
      new StringPromptTemplate({
        template: template,
        inputVariables: inputVariables,
      });
    }).toThrow(
      "Variable 'unusedVariable' is declared but not used in the template."
    );
  });

  test('adding guardrails', async () => {
    const isStringRule = new GeneralRule({
      description: 'is string',
      func: async (input: unknown) => {
        return typeof input === 'string';
      }
    });

    const promptTemplate = new StringPromptTemplate({
      template: '{{name}}',
      inputVariables: ['name'],
      guardrails: {
        default: isStringRule
      }
    });

    expect(() => promptTemplate.invoke({ name: 1 })).rejects.toThrow(
      "CANNOT format prompt because of error - Validation failed in variable 'name'"
    );
  });

  test('prefix and suffix in the template', async () => {
    const promptTemplate = new StringPromptTemplate({
      template: '{{name}}',
      inputVariables: ['name'],

      prefix: 'Hello, ',
      suffix: '!',
    });

    const formattedString = await promptTemplate.invoke({ name: 'Alice' });
    expect(formattedString.toString()).toBe('Hello, Alice!');
  });
});
