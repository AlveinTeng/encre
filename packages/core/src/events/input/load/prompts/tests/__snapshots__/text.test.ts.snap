// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test custom StringPrompt 1`] = `
"_grp: 1
_type: constructor
_id:
  - events
  - input
  - load
  - prompts
  - text
  - StringPrompt
_kwargs:
  value: this is a test prompt.
"
`;

exports[`test custom StringPrompt 2`] = `
"- _grp: 1
  _type: constructor
  _id:
    - events
    - input
    - load
    - msgs
    - human
    - HumanMessage
  _kwargs:
    content: this is a test prompt.
"
`;
